### PR TITLE
oncoprint - add support for categorical generic assay tracks

### PR DESCRIPTION
**Pull request by adamabeshouse**
Fixes https://github.com/cBioPortal/cbioportal/issues/8313

image

**Test instruction:** 
1. Go to study view page: http://private.cbioportal.org/private/study/summary?id=acc_tcga_pan_can_atlas_2018
2. Set `localStorage.netlify ="deploy-preview-3711--cbioportalfrontend"` in browser console.
3. Query any gene, and go to result view page
4. Select OncoPrint tab
5. Click `Add Tracks`, then add tracks from `Armlevel Cna` tab

### CODE REVIEW

File: src/shared/lib/onMobxPromise.ts [hunk 1]
```diff
@@ -29,4 +29,10 @@ export function onMobxPromise<T>(
     return disposer;
 }
 
+export function toPromise<T>(promise: MobxPromise<T>): Promise<T> {
```
Comment (alisman) [inline src/shared/lib/onMobxPromise.ts:32]: curious if this handles error

Comment (adamabeshouse) [reply src/shared/lib/onMobxPromise.ts:32]: not gracefully

File: src/shared/components/oncoprint/DataUtils.ts [hunk 2]
```diff
@@ -487,6 +489,77 @@ function selectRepresentingDataPoint(
...
+
+export function makeCategoricalTrackData(
+    profile: MolecularProfile,
+    entityId: string,
+    cases: Sample[] | Patient[],
+    data: GenericAssayData[]
+) {
+    let ret: CategoricalTrackDatum[];
+    let keyToData: { [uniqueKey: string]: GenericAssayData[] };
+    if (isSampleList(cases)) {
```
Comment (alisman) [inline src/shared/components/oncoprint/DataUtils.ts:525]: maybe put a comment here explaining sample vs case list fork?

File: src/shared/components/oncoprint/DeltaUtils.ts [hunk 3]
```diff
@@ -102,10 +106,13 @@ function transitionTrackHeaders(
     oncoprint: OncoprintJS
 ) {
     if (
-        !_.isEqual(nextProps.heatmapTrackHeaders, prevProps.heatmapTrackHeaders)
+        !_.isEqual(
```
Comment (alisman) [inline src/shared/components/oncoprint/DeltaUtils.ts:109]: i know this is old and venerable logic but it would help to have a comment explaining what "track transition" is in layman's terms.

File: end-to-end-test/local/specs/core/oncoprint.screenshot.spec.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+var assert = require('assert');
```
Comment (dippindots) [inline end-to-end-test/local/specs/core/oncoprint.screenshot.spec.js:1]: assert seems has not been used at here.

File: src/shared/components/oncoprint/ResultsViewOncoprint.tsx [hunk 5]
```diff
@@ -1310,6 +1370,9 @@ export default class ResultsViewOncoprint extends React.Component<
                     ),
                     ...this.genericAssayHeatmapTracks.result.map(
                         hmTrack => hmTrack.trackGroupIndex
+                    ),
+                    ...this.genericAssayCategoricalTracks.result.map(
+                        track => track.trackGroupIndex
```
Comment (dippindots) [inline src/shared/components/oncoprint/ResultsViewOncoprint.tsx:1375]: Do we want to add `CategoricalTracks` in `HeatmapTrackGroup`? Not sure if this is because we are using `HeatmapTrack` to represent `AdditionalTrack`? If so, maybe we should rename some of the variables.

File: src/shared/components/oncoprint/ResultsViewOncoprint.tsx [hunk 6]
```diff
@@ -1349,7 +1412,7 @@ export default class ResultsViewOncoprint extends React.Component<
             if (clusteredHeatmapProfile === genesetHeatmapProfile) {
                 return this.genesetHeatmapTrackGroupIndex;
             } else {
-                const heatmapGroup = this.molecularProfileIdToHeatmapTracks[
+                const heatmapGroup = this.molecularProfileIdToAdditionalTracks[
```
Comment (dippindots) [inline src/shared/components/oncoprint/ResultsViewOncoprint.tsx:1415]: Same here,  if we are using `HeatmapTrack` to represent `AdditionalTrack`, maybe we should rename some of the variables.

File: src/shared/components/oncoprint/TooltipUtils.ts [hunk 7]
```diff
@@ -200,7 +261,10 @@ export function makeClinicalTrackTooltip(
     };
 }
 export function makeHeatmapTrackTooltip(
-    trackSpec: IBaseHeatmapTrackSpec,
+    trackSpec: Pick<
+        IBaseHeatmapTrackSpec,
+        'tooltipValueLabel' | 'molecularAlterationType'
```
Comment (dippindots) [inline src/shared/components/oncoprint/TooltipUtils.ts:266]: Is this `tooltipValueLabel` necessary?

**Review by dippindots — COMMENTED** [review]
Thanks for making this happen!!

Overall it looks good to me, I just had a few comments about naming.

**Comment by dippindots** [comment]
@adamabeshouse I added `Test instruction` in the description for product team to test this feature. Please feel free to change it anytime!

**Comment by jjgao** [comment]
Looks great!

- [ ] Change the order in Plots to (lost, diploid, gained, NA)
- [ ] Use the `not profiled` glyph for NA?

File: src/shared/components/oncoprint/OncoprintUtils.ts [hunk 8]
```diff
@@ -1041,19 +1053,19 @@ export function makeHeatmapTracksMobxPromise(
...
+                            oncoprint.molecularProfileIdToAdditionalTracks[
                                 molecularProfileId
                             ];
                         if (trackGroup) {
                             const newEntities = _.keys(
                                 trackGroup.entities
                             ).filter(entity => entity !== gene);
                             if (newEntities.length === 0) {
-                                oncoprint.removeHeatmapByMolecularProfileId(
+                                oncoprint.removeHeatmapTracksByMolecularProfileId(
```
[the remark this answers is not in the document]
Comment (adamabeshouse) [reply src/shared/components/oncoprint/OncoprintUtils.ts:1068]: no this is specifically for heatmap tracks. there is a separate method for categorical tracks, since they are handled differently

File: src/shared/components/oncoprint/OncoprintUtils.ts [hunk 9]
```diff
@@ -1159,6 +1171,144 @@ export function makeGenericAssayProfileHeatmapTracksMobxPromise(
...
+                    .genericAssayMolecularDataCache.result!;
+
+                const entityId = query.stableId;
+                const genericAssayType = profile.genericAssayType;
+                const entityLinkMap = oncoprint.props.store
+                    .genericAssayEntitiesGroupedByGenericAssayTypeLinkMap
+                    .result![profile.genericAssayType];
+
+                return {
+                    key: `GENERICASSAYHEATMAPTRACK_${molecularProfileId},${entityId}`,
```
[the remark this answers is not in the document]
Comment (adamabeshouse) [reply src/shared/components/oncoprint/OncoprintUtils.ts:1188]: whoa good catch

File: src/shared/components/oncoprint/ResultsViewOncoprint.tsx (hunk 5 above)
Comment (adamabeshouse) [reply src/shared/components/oncoprint/ResultsViewOncoprint.tsx:1375]: this is just computing what group index will be the geneset heatmaps. this just mean sit will go all the way at the end after heatmap tracks, generic assay heatmap tracks, and generic assay categorical tracks

File: src/shared/components/oncoprint/ResultsViewOncoprint.tsx (hunk 6 above)
Comment (adamabeshouse) [reply src/shared/components/oncoprint/ResultsViewOncoprint.tsx:1415]: additionaltracks holds all of these, and because clusteredHeatmapProfile we know is heatmap, we know that this entry is a heatmap group. I'm not sure I understand the issue

File: src/shared/components/oncoprint/TooltipUtils.ts (hunk 7 above)
Comment (adamabeshouse) [reply src/shared/components/oncoprint/TooltipUtils.ts:266]: yeah, its used just below, right?

File: src/shared/components/oncoprint/OncoprintUtils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (dippindots) [reply src/shared/components/oncoprint/OncoprintUtils.ts:1068]: got it thanks!

File: src/shared/components/oncoprint/TooltipUtils.ts (hunk 7 above)
Comment (dippindots) [reply src/shared/components/oncoprint/TooltipUtils.ts:266]: Sorry I overlooked that part, thanks

**Comment by dippindots** [comment]
@adamabeshouse I addressed some comments in my pr, and here can be a reference of `category order`: https://github.com/cBioPortal/cbioportal-frontend/pull/3694/files#diff-b99a0de889ebf6dd268cd774d552550ce6d40c0bf77e09a9ce400d7bde19766fR244-R250

I will fix the `NA` data issue in backend, maybe we can hold on a while on JJ's second comments, or if you think it's not data related, please go ahead with your idea.

And please see Tali's comments at here https://github.com/cBioPortal/cbioportal/issues/8601. I will fix the first one, and seems 4-6 will be fixed if we fixed the heatmap issue.

Thanks again for your help! Appreciated!

**Comment by adamabeshouse** [comment]
@dippindots thanks! I fixed the heatmap issue, let's talk when you get back about whats left
